compiler/rustc_session/src/config.rs

```diff
@@ -1005,6 +1006,96 @@ pub fn to_crate_config(cfg: FxHashSet<(String, Option<String>)>) -> CrateConfig
     cfg.into_iter().map(|(a, b)| (Symbol::intern(&a), b.map(|b| Symbol::intern(&b)))).collect()
 }
 
+/// The parsed `--check-cfg` options
+pub struct CheckCfg<T = String> {
+    /// Set if `names()` checking is enabled
+    pub names_checked: bool,
+    /// The union of all `names()`
+    pub names_valid: FxHashSet<T>,
+    /// The set of names for which `values()` was used
+    pub values_checked: FxHashSet<T>,
+    /// The set of all (name, value) pairs passed in `values()`
+    pub values_valid: FxHashSet<(T, T)>,
+}
+
+impl<T> Default for CheckCfg<T> {
+    fn default() -> Self {
+        CheckCfg {
+            names_checked: false,
+            names_valid: FxHashSet::default(),
+            values_checked: FxHashSet::default(),
+            values_valid: FxHashSet::default(),
+        }
+    }
+}
+
+impl<T> CheckCfg<T> {
+    fn map_data<O, F>(&self, f: F) -> CheckCfg<O>
+    where
+        F: Fn(&T) -> O,
+        O: std::cmp::Eq + std::hash::Hash,
+    {
+        CheckCfg::<O> {
+            names_checked: self.names_checked,
+            names_valid: (&self.names_valid).into_iter().map(|a| f(a)).collect(),
+            values_checked: (&self.values_checked).into_iter().map(|a| f(a)).collect(),
+            values_valid: (&self.values_valid).into_iter().map(|(a, b)| (f(a), f(b))).collect(),
+        }
+    }
+}
+
+/// Converts the crate `--check-cfg` options from `String` to `Symbol`.
+/// `rustc_interface::interface::Config` accepts this in the compiler configuration,
+/// but the symbol interner is not yet set up then, so we must convert it later.
+pub fn to_check_config(cfg: CheckCfg) -> CrateCheckConfig {
```

Comment:
This function can be refactored away now.

Reply:
I'm would prefer not, I find weird to have this kind of call `check_cfg.map_data(|a| Symbol::intern(a));` in the middle of a code; it's not explanatory (it requires the reader to be aware of the implementation details of `CrateCheckConfig`) instead of just having a call to a function that is self explanatory; and it's also not done for [`to_crate_config`](https://github.com/rust-lang/rust/blob/a240ccd81c74c105b6f5fe84c46f8d36edb7e306/compiler/rustc_session/src/config.rs#L1016-L1021) (the equivalent for `cfg`s).

Reply:
Ok, could you at least rename it to `to_crate_check_config` then, to match naming of the output type?
And perhaps turn it into a method on `CheckCfg`.

And the functions below (`fill_check_config_well_known`, `fill_check_config_actual`) can perhaps be turned into methods on `CrateCheckConfig` too.

Reply:
Done.